dali/kernels/imgproc/color_manipulation/hsv_cpu.h

```diff
@@ -0,0 +1,139 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_IMGPROC_COLOR_MANIPULATION_HSV_CPU_H_
+#define DALI_KERNELS_IMGPROC_COLOR_MANIPULATION_HSV_CPU_H_
+
+#include <utility>
+#include "dali/util/ocv.h"
+#include "dali/core/common.h"
+#include "dali/core/convert.h"
+#include "dali/core/geom/box.h"
+#include "dali/core/error_handling.h"
+#include "dali/kernels/kernel.h"
+#include "dali/pipeline/data/types.h"
+
+namespace dali {
+namespace kernels {
+
+namespace hsv {
+
+constexpr size_t ndims = 3;
+constexpr size_t nchannels = 3;
+
+/**
+ * Defines region of interest.
+ * 0 dimension is interpreted along x axis (horizontal)
+ * 1 dimension is interpreted along y axis (vertical)
+ *
+ *            image.x ->
+ *          +--------------------------------+
+ *          |        roi.x                   |
+ *  image.y |      +-----+                   |
+ *       |  | roi.y|     |                   |
+ *       v  |      +-----+                   |
+ *          +--------------------------------+
+ */
+template<size_t ndims>
+using Roi = Box<ndims, int>;
+
+
+/**
+ * Defines TensorShape corresponding to provided Roi.
+ * Assumes HWC memory layout
+ *
+ * @tparam ndims_roi Number of dims in Roi
+ * @param roi Region of interest
+ * @param nchannels Number of channels in data
+ * @return Corresponding TensorShape
+ */
+template<size_t ndims_roi>
+TensorShape<ndims_roi + 1> roi_shape(Roi<ndims_roi> roi, size_t nchannels) {
+    assert(all_coords(roi.hi >= roi.lo) && "Cannot create a tensor shape from an invalid Box");
+    TensorShape<ndims_roi + 1> ret;
+    auto e = roi.extent();
+    auto ridx = ndims_roi;
+    ret[ridx--] = nchannels;
+    for (size_t idx = 0; idx < ndims_roi; idx++) {
+        ret[ridx--] = e[idx];
+    }
+    return ret;
+}
+
+}  // namespace hsv
+
+
+template<class OutputType, class InputType>
+class HsvCpu {
+ public:
+    using Roi = hsv::Roi<hsv::ndims - 1>;
+
+
+    KernelRequirements
+    Setup(KernelContext &context, const InTensorCPU<InputType, hsv::ndims> &in, float hue,
+          float saturation, float value, const Roi *roi = nullptr) {
+        DALI_ENFORCE(!roi || all_coords(roi->hi >= roi->lo), "Region of interest is invalid");
+        auto adjusted_roi = AdjustRoi(roi, in.shape);
```

Comment:
Not really, because this function also takes care of the case, where ROI is not provided (then, returned Roi is of the size of the image). So it's not really clipping anything, rather expanding